Add hyperelliptic curves using the smooth model

GiacomoPope opened the pull request:
## Overview

This PR includes a whole new module into schemes which implements hyperelliptic curves over the smooth model, and focuses on implementing decent arithmetic for Jac(H) for all reasonable cases, something which is not done properly in the current implementation due to limitations of the curve model used.

The idea is that `hyperelliptic_curves_sm` will be available with the old model for some time, until someone decides we should depreciate the old model in favour of this new implementation.

## Motivation

The current implementation of hyperelliptic curves in SageMath uses the projective plane model. Although this works nicely enough for imaginary curves with only one point at infinity, it is not descriptive enough for the real models. My gut feeling is the projective plane model was used as in early Sage days this was easier to do and allowed hyperelliptic curves to be supported at all. Mathematically, I believe it makes much more sense to use a weighted projective model to have this new description but it requires more tools which we only have thanks to the work of others in sage since it was released.

This PR is a total rewrite of the hyperelliptic curve classes to instead use the smooth model for hyperelliptic curves which facilitates implementing arithmetic of Jacobians of hyperelliptic curves for (almost) all cases. In particular, now if one tries to perform arithmetic on Jac(C) one either gets the correct value or a well-handled error instead of just returning something which is wrong.

The hope is that with this new model for the curves, more recent research in the area of hyperelliptic curves and their jacobians can be more easily integrated. As a personal example, being able to compute arithmetic in Jac(H) for the real model unconditionally would help with the implementation of genus two isogenies which are "in vogue" right now in the cryptography world. 

### Content of PR

This PR looks WAY bigger than it is, as it duplicates ALL code from `hyperelliptic_curves` and then has modifications to allow this to work in the new curve model. I believe the best thing to do in the long run is to depreciate the old hyperelliptic curve impl and have this replace it, but this will take years(?) so I think we'll just have to have both side by side for a while.

This PR still needs a lot of work including:

- more tests
- better comments and docstrings
- potentially refactoring of which files belong where

but the code has sat at this stage for a long time without much more progress from the three of us, so I think the best thing to do is open up the PR and try and get some extra attention on getting this code ready to be included.

One benefit is that all this code is "new" in that it should not conflict with anyone else's work, so we can get this code into a good position and keep adding to it with more interesting maths once the base layer is in place

### Example of better arithmetic

For example, this code fixed the following issue:

https://github.com/sagemath/sage/issues/32024

```py
sage: R.<x> = QQ[]
sage: f = 144*x^6 - 240*x^5 + 148*x^4 + 16*x^3 - 16*x^2 - 4*x + 1
sage: H = HyperellipticCurveSmoothModel(f)
sage: J = Jacobian(H)
sage: P = J(H(0,1))-J(H(0,-1))
sage: (5*P).is_zero()
False
sage: 
sage: H = HyperellipticCurve(f)
sage: J = Jacobian(H)
sage: P = J(H(0,1))-J(H(0,-1))
sage: (5*P).is_zero()
True
```

and is also related to the issues https://github.com/sagemath/sage/issues/37109 https://github.com/sagemath/sage/issues/37093 https://github.com/sagemath/sage/issues/37101 https://github.com/sagemath/sage/issues/37626

### Help!

This code needs more work, and lots of time spent on the review which I don't think is a reasonable thing to do for one person. So if this area interests you any help would be appreciated (either in the review of some of the code or some extra commits which tidy up aspects)

GiacomoPope committed:
start porting code ready to be integrated into sagemath

Co-Authored-By: grhkm21 <83517584+grhkm21@users.noreply.github.com>
Co-Authored-By: sabrinakunzweiler <22136626+sabrinakunzweiler@users.noreply.github.com>

GiacomoPope committed:
start work on porting to sage

GiacomoPope committed:
fix import lines

GiacomoPope committed:
fix function imports

GiacomoPope committed:
fix a bunch of old import statements

grhkm21 committed:
fix some monsky_washnitzer imports

grhkm21 committed:
fix all doctest errors

GiacomoPope committed:
add long time warning to doctest

GiacomoPope committed:
run linting checks

GiacomoPope committed:
remove is_HyperellipticCurveSmoothModel as it's not really a modern solution

GiacomoPope committed:
some proof reading

GiacomoPope committed:
fix stupid bug

grhkm21 committed:
Merge commit '5a8daf4e29c0124bcfc4cd0758de4a81f47b1fc6' into hyperelliptic-curve

grhkm21 committed:
Merge remote-tracking branch 'refs/remotes/jack/hyperelliptic_curve_sm' into hyperelliptic-curve

grhkm21 committed:
Merge remote-tracking branch 'origin/develop' into hyperelliptic_curve_sm

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

GiacomoPope committed:
pycodestyle fix

GiacomoPope committed:
fix some rst issues, some remain...

GiacomoPope committed:
try and fix more RST issues

grhkm21 committed:
Merge remote-tracking branch 'refs/remotes/jack/hyperelliptic_curve_sm' into hyperelliptic_curve_sm

grhkm21 committed:
Moved weighted files and changed imports

grhkm21 committed:
remove random file...

grhkm21 committed:
remove more debugs...

GiacomoPope committed:
more RST

GiacomoPope committed:
fix failing doctests

GiacomoPope committed:
fix typing error

grhkm21 committed:
implement comparison between weighted projective points

grhkm21 committed:
implement zero divisor check for constructor of points in weighted projective space

grhkm21 committed:
reimplement check more directly

grhkm21 committed:
remove extra import

grhkm21 committed:
move weighted_projective_curve to schemes/curves

grhkm21 committed:
update meson.build

grhkm21 committed:
sort imports and remove extra export

grhkm21 committed:
add back missing import for WeightedProjectiveSpace

grhkm21 committed:
run tools/update_meson.py (also done in #39159)

grhkm21 committed:
fix doctest errors 2

grhkm21 committed:
(TODO?) fix normalisation of points in weighted projective space

We chose not to normalise all points, but rather just the points where
the weight of the coordinate corresponding to the last nonzero entry of
the points

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

sabrinakunzweiler committed:
fix and add some docstrings

sabrinakunzweiler committed:
include our files in the documentation and fix some docstrings

fchapoton commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_g2.py:
```diff
@@ -76,32 +78,28 @@ def jacobian(self):
 
         The points `P = (0, 0)` and `Q = (-1, -1)` are on `H`. We construct the
-        element `D_1 = [P - Q] = [P + (-Q) - D_\infty`] on the Jacobian::
+        element `D_1 = [P - Q] = [P + (-Q) - D_\infty]` on the Jacobian.
```
all the changes about removing `::` are bad and should be reverted

sabrinakunzweiler replied:
done

sabrinakunzweiler committed:
fix docstrings again

fchapoton commented on src/sage/schemes/hyperelliptic_curves_smooth_model/jacobian_homset_split.py:
```diff
@@ -0,0 +1,349 @@
+        return self._morphism_element(self, u, v, n=n, check=check)
+
+    def cantor_composition(self, u1, v1, n1, u2, v2, n2):
+        """
```
```suggestion
        r"""
```

fchapoton replied:
This will fix the warning about `\i`

sabrinakunzweiler committed:
Update src/sage/schemes/hyperelliptic_curves_smooth_model/jacobian_homset_split.py

Co-authored-by: Frédéric Chapoton <fchapoton2@gmail.com>

fchapoton commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,262 @@
+        <class 'sage.schemes.hyperelliptic_curves_smooth_model.hyperelliptic_padic_field.HyperellipticCurveSmoothModel_padic_field_with_category'>
+
+    The input polynomials need not be monic::
+    
```
```suggestion

```

fchapoton replied:
see the Lint warnings about spaces at the end of lines : they should be removed

fchapoton commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_g2.py:
```diff
@@ -0,0 +1,278 @@
+    @cached_method
+    def jacobian(self):
+        r"""
+        Returns the Jacobian of the hyperelliptic curve.
```
```suggestion
        Return the Jacobian of the hyperelliptic curve.
```

fchapoton replied:
The first line in the doc should use "imperative mode", like when giving an order.

sabrinakunzweiler committed:
Update src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py

Co-authored-by: Frédéric Chapoton <fchapoton2@gmail.com>

sabrinakunzweiler committed:
Update src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_g2.py

Co-authored-by: Frédéric Chapoton <fchapoton2@gmail.com>

sabrinakunzweiler committed:
delete whitespace

sabrinakunzweiler committed:
add more docstrings

sabrinakunzweiler committed:
minor

grhkm21 committed:
Merge branch 'weighted-projective' into hyperelliptic_curve_sm

grhkm21 committed:
Merge remote-tracking branch 'refs/remotes/jack/hyperelliptic_curve_sm' into hyperelliptic_curve_sm

grhkm21 committed:
add additional tests for points_at_infinity

grhkm21 committed:
fix doctest errors

grhkm21 committed:
Merge branch 'weighted-projective' into hyperelliptic_curve_sm

grhkm21 committed:
implement discriminant formula for hyperelliptic curves

grhkm21 committed:
J.order only works over finite fields

grhkm21 committed:
fix issues with coerce_map_from not firing for polynomial rings

grhkm21 committed:
add `distinguished_point` to hyperelliptic curve constructor

grhkm21 committed:
Merge remote-tracking branch 'sagemath/develop' into hyperelliptic_curve_sm

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

fchapoton committed:
Update weighted_projective_space.py

partial fix for the linter

GiacomoPope committed:
fix RST lint issue

GiacomoPope committed:
fix whitespace linter issue

GiacomoPope committed:
fix doctests

GiacomoPope committed:
add authors and copyright

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+in weighted projective space `\mathbb{P}(1 : g + 1 : 1)`.
+
+Adapted from /hyperelliptic/constructor.py
+
```
```suggestion
Adapted from ``sage/schemes/hyperelliptic_curves/constructor.py``
```

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/constructor.py:
```diff
@@ -0,0 +1,279 @@
+r"""
```
Maybe just name this file ``constructor.py`` for consistency with the existing elliptic and hyperelliptic packages?

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+# ****************************************************************************
+#       Copyright (C) 2025 Sabrina Kunzweiler, Gareth Ma, Giacomo Pope
+#                     2019 Anna Somoza <anna.somoza.henares@gmail.com>
+#                     2006 David Kohel <kohel@maths.usyd.edu>
```
```suggestion
#       Copyright (C) 2006 David Kohel <kohel@maths.usyd.edu>
#                     2019 Anna Somoza <anna.somoza.henares@gmail.com>
#                     2025 Sabrina Kunzweiler
#                     2025 Gareth Ma
#                     2025 Giacomo Pope

```

This should be sorted from oldest to newest. For the sake of consistency, I think it's better to have one person per line. I think it is encouraged but not required to list the email addresses of the authors as well.

This applies to all the copyright headers.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+    HyperellipticCurveSmoothModel_rational_field,
+)
+
+"""
```
Move this into the main docstring for the module so it shows up in the documentation.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+
+
+def HyperellipticCurveSmoothModel(
+    f, h=0, check_squarefree=True, distinguished_point=None
```
```suggestion
    f, h=0, check_squarefree: bool = True, distinguished_point=None
```

While we don't require type annotations, they are encouraged. Especially when you're dealing with basic types where it's easy to figure out what the correct annotation is.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/meson.build:
```diff
@@ -0,0 +1,27 @@
+  subdir: 'sage/schemes/hyperelliptic_curves_smooth_model',
+)
+
+extension_data_cpp = {}
```
I don't think this should be here if you aren't using C++ in any Cython files.

Did you create this file manually? We have a script `tools/update-meson.py` that creates the `meson.build` files (mostly) automatically. Unless you're doing something exotic that script should "Just Work".

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/jacobian_generic.py:
```diff
@@ -0,0 +1,196 @@
+
+        return self.point_homset().points()
+
+    def __iter__(self):
```
The benefit of `__iter__` is that you don't need to construct the entire list before you can start doing computations with elements. The way you've implemented this of yielding from an already-constructed list means we don't get that benefit. Ideally this should be implemented the other way around, where `__iter__` is a something like:

```python
def __iter__(self) -> Iterator:
    for foo in bar:
        yield baz(foo)
```

and then `list` is just `return list(iter(self))`.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+    # Internal Helper functions
+    # ---------------------------
+
+    def __genus(f, h):
```
```suggestion
    def genus(f, h) -> Integer:
```

The reason we use (single) underscores is to hide a function/method from the public API. Helper functions that are internal to another function aren't in the public API anyway, so we don't prefix their names with underscores.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/constructor.py:
```diff
@@ -0,0 +1,279 @@
+
+        sage: K.<alpha> = QQ.extension(x^2+x-1)
+        sage: HK = H.change_ring(K)
+        sage: HK.points_at_infinity()
```
What's going on with these test failures?

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+        dh_2 = 2 * Integer(h.degree())
+        if dh_2 < df:
+            return (df - 1) // 2
+        return (dh_2 - 1) // 2
```
```suggestion
        # Some classes still have issues with degrees returning `int`
        # rather than Sage Integer types
        df = f.degree()
        dh_2 = 2 * h.degree()
        if dh_2 < df:
            return Integer((df - 1) // 2)
        return Integer((dh_2 - 1) // 2)
```

vincentmacri replied:
This version only does one `int` to `Integer` conversion instead of two.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/constructor.py:
```diff
@@ -0,0 +1,279 @@
+            return (df - 1) // 2
+        return (dh_2 - 1) // 2
+
+    def __check_no_affine_singularities(f, h):
```
```suggestion
    def check_no_affine_singularities(f, h) -> bool:
```

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+        return (dh_2 - 1) // 2
+
+    def __check_no_affine_singularities(f, h):
+        """
```
```suggestion
        r"""
```

It technically doesn't matter here since these functions don't show up in the docs, but using `r"""` instead of `"""` reduces the amount of escape sequences needed for some LaTeX expressions, and so we prefer it.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+            H = sum(h[i] * X**i * Z ** (d - i) for i in range(d + 1))
+            G = Y**2 + H * Y - F
+
+        return G
```
```suggestion
        if h.is_zero():
            return Y**2 - F
        H = sum(h[i] * X**i * Z ** (d - i) for i in range(d + 1))
        return Y**2 + H * Y - F
```

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/constructor.py:
```diff
@@ -0,0 +1,279 @@
+        Helper function to compute the genus of a hyperelliptic curve
+        defined by `y^2 + h(x)y = f(x)`.
+        """
+        # Some classes still have issues with degrees returning `int`
```
If you know which classes these are, you can fix them in this PR. If not, it's fine.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_g2.py:
```diff
@@ -0,0 +1,293 @@
+
+class HyperellipticCurveSmoothModel_g2(
+    hyperelliptic_generic.HyperellipticCurveSmoothModel_generic
+):
```
```suggestion
from sage.schemes.hyperelliptic_curves_smooth_model import (
    invariants,
)
from sage.schemes.hyperelliptic_curves_smooth_model.hyperelliptic_finite_field import (
    HyperellipticCurveSmoothModel_finite_field,
)
from sage.schemes.hyperelliptic_curves_smooth_model.hyperelliptic_padic_field import (
    HyperellipticCurveSmoothModel_padic_field,
)
from sage.schemes.hyperelliptic_curves_smooth_model.hyperelliptic_rational_field import (
    HyperellipticCurveSmoothModel_rational_field,
)

from .hyperelliptic_generic import HyperellipticCurveSmoothModel_generic

"""
TODO List

### Invariants

There seems to be massive redundancy in having these methods and the ones imported into
invariants. I think we should fix this by putting the methods themseleves into this class.
"""


class HyperellipticCurveSmoothModel_g2(HyperellipticCurveSmoothModel_generic):
```

You should be able to simplify the other imports similarly.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+        X, Y, Z = PolynomialRing(f.base_ring(), names="X, Y, Z").gens()
+
+        # Some classes still have issues with degrees returning `int`
+        d = max(Integer(h.degree()), (Integer(f.degree()) + 1) // 2)
```
```suggestion
        d = Integer(max(h.degree(), (f.degree() + 1) // 2))
```

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py:
```diff
@@ -0,0 +1,279 @@
+        dh_2 = 2 * Integer(h.degree())
+        if dh_2 < df:
+            return (df - 1) // 2
+        return (dh_2 - 1) // 2
```
I feel like I should know this, but my flight was at an awful hour and I don't sleep well on planes so my brain is not running at 100%. Can you remind me where this formula comes from? What's it called? (And maybe add an inline comment with the name of the formula so people reading the code in the future can look it up.) Same comment applies for a lot of the formulas you're using here, an inline comment saying something like `# Bob's formula` or `# Sally's theorem` would help a lot with readability.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/constructor.py:
```diff
@@ -0,0 +1,279 @@
+
+    # Check the polynomials are of the right type
+    F = h**2 + 4 * f
+    if not isinstance(F, Polynomial):
```
Can we not annotate the input parameters with `Polynomial` as well? That doesn't mean you can remove this check, since type annotations are not checked at runtime, but it does let our CI catch any violations within library code, and it also makes things a bit more clear to users reading the documentation.

vincentmacri reviewed (changes requested):
A quick first look, I only got a chance to look at some of the files so this is by no means exhaustive feedback. But there is some general formatting/style/best practices feedback that you can apply throughout. I have some more general questions that I'll post in a separate comment.

With other things I had going on mostly out of the way now (i.e. most of my PhD coursework) this PR is now near the top of my priority list, so other than the upcoming holidays and my conference travel this week I hope to be able to respond fairly quickly now.

vincentmacri commented on src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_g2.py:
```diff
@@ -0,0 +1,293 @@
+
+        from sage.schemes.hyperelliptic_curves_smooth_model.jacobian_g2_generic import (
+            HyperellipticJacobian_g2_generic,
+        )
```
It's better to have all imports at the top of the file at the module level unless they are expensive or cause circular import issues.

GiacomoPope committed:
Update src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

GiacomoPope committed:
change copyright

GiacomoPope committed:
reviewer feedback

GiacomoPope committed:
add r to doctrings

GiacomoPope committed:
reviewer feedback

GiacomoPope committed:
reviewer feedback

GiacomoPope committed:
try working in some deprcation

GiacomoPope committed:
Update constructor.py

GiacomoPope committed:
Update constructor.py

GiacomoPope committed:
Update src/sage/schemes/hyperelliptic_curves/constructor.py

Co-authored-by: Frédéric Chapoton <chapoton@unistra.fr>

GiacomoPope committed:
lint fix

vincentmacri reviewed (commented):
Try to fix the test failures.

The fact that a lot of this was copied from `hyperelliptic_curves` to `hyperelliptic_curves_smooth_model` makes it hard to see what was changed. Is there a commit in the history where you copied `hyperelliptic_curves` to `hyperelliptic_curves_smooth_model` but didn't make any changes yet, so I can diff between those commits when looking at this? If not, would you mind creating such a commit (either do some rebase magic, or delete all files in `hyperelliptic_curves_smooth_model`, copy all files in `hyperelliptic_curves` to `hyperelliptic_curves_smooth_model`, commit, then undo it and commit again).

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

GiacomoPope committed:
revert depreciation warnings and fix some tests

GiacomoPope committed:
remove circular import

GiacomoPope committed:
move code from smooth model to replacing the old model, some code not passing tests

vincentmacri reviewed (changes requested):
First review pass of `src/sage/schemes/hyperelliptic_curves/constructor.py` (and a couple comments on other files).

Looks good, mostly just code style nitpicks and documentation suggestions.

vincentmacri reviewed (changes requested):
First pass of hyperelliptic_generic.

GiacomoPope committed:
use lazy import for circular import

GiacomoPope committed:
reviewer feedback, some tests still fail

GiacomoPope committed:
Apply suggestions from code review

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

GiacomoPope committed:
small fixes after batch commiting reviewer feedback

sabrinakunzweiler committed:
fix description of homset over extensions

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

GiacomoPope committed:
fix ruff failures

GiacomoPope committed:
move _parse_multivariate_defining_equation into file where it's used

GiacomoPope committed:
add printing ring back to the construtor

GiacomoPope committed:
remove trailing whitespace

sabrinakunzweiler committed:
Merge branch 'hyperelliptic_curve_sm' of https://github.com/GiacomoPope/sage into hyperelliptic_curve_sm

GiacomoPope committed:
fix import for doctest

GiacomoPope committed:
Apply suggestions from code review

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

sabrinakunzweiler committed:
add kummer surface functionality

sabrinakunzweiler committed:
Merge branch 'hyperelliptic_curve_sm' of https://github.com/GiacomoPope/sage into hyperelliptic_curve_sm

GiacomoPope committed:
run formatter

GiacomoPope committed:
fix doctests

GiacomoPope committed:
add basic functionality for reimann surface for weighted projective space

GiacomoPope committed:
fix additional doctests

GiacomoPope committed:
fix meson build file

vincentmacri reviewed (changes requested):
Looks good so far, most requested changes are just small formatting things that you can accept in the GitHub UI. I haven't taken a detailed look at the correctness of the Jacobian arithmetic implementation yet.

You should try to add a doctest that uses `TestSuite.run` on all objects you've added, especially for groups or other objects where Sage can check that the axioms hold for some random elements.

GiacomoPope committed:
Apply suggestions from code review

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

GiacomoPope committed:
swap shift for exp

GiacomoPope committed:
change docstring to make computations for EllipticCurve_from_Weierstrass_polynomial more obvious

GiacomoPope committed:
use TYPE_CHECKING for integer

GiacomoPope committed:
fix import statements for typechecking

GiacomoPope committed:
partial coverage with test suite

grhkm21 committed:
pickle/unpickle for Jacobian elements, and a bunch of stuff

grhkm21 committed:
Merge remote-tracking branch 'refs/remotes/jack/hyperelliptic_curve_sm' into hyperelliptic-smooth

vincentmacri committed:
Do not use __call__

Add str annotation

Avoid use of __call__

Better some_elements

Cleanup

Linter fix

Formatting

vincentmacri reviewed (changes requested):
The diffs are too long to easily read in my browser. Is it alright if I submit the rest of my suggestions via a PR to your branch?

GiacomoPope committed:
Merge pull request #2 from vincentmacri/hyperelliptic-suggestions

Hyperelliptic suggestions

vincentmacri committed:
Make u monic earlier, making it less likely to be forgotten.

Fix and add references

Some cleanup and notes

GiacomoPope committed:
Merge branch 'develop' into hyperelliptic_curve_sm

GiacomoPope committed:
Merge pull request #3 from vincentmacri/hyperelliptic-suggestions

Hyperelliptic suggestions

GiacomoPope committed:
fix issues arising from ruff check --preview --fix --select RET src/sage/schemes/hyperelliptic_curves

GiacomoPope committed:
run ruff check --preview --fix src/sage/schemes/hyperelliptic_curves
